tests/src/Kernel/Entity/GetUserMembershipsTest.php

```diff
@@ -0,0 +1,229 @@
+      [NULL, NULL],
+    ];
+
+    foreach ($matrix as $user_key => $statuses) {
+      $user = User::create(['name' => $this->randomString()]);
+      $user->save();
+      $this->users[$user_key] = $user;
+      foreach ($statuses as $group_key => $status) {
+        $group = $this->groups[$group_key];
+        if ($status) {
+          $membership = OgMembership::create(['type' => OgMembershipInterface::TYPE_DEFAULT]);
+          $membership
+            ->setUser($user->id())
+            ->setEntityId($group->id())
+            ->setGroupEntityType($group->getEntityTypeId())
+            ->setState($status)
+            ->save();
+        }
+      }
+    }
+  }
+
+  /**
+   * Tests retrieval of OG Membership entities associated with a given user.
+   *
+   * @param int $user
+   *   The key of the user in the $this->users array.
+   * @param array $states
+   *   Array with the states to retrieve.
+   * @param string $field_name
+   *   The field name associated with the group.
+   * @param array $expected
+   *   An array containing the expected results to be returned.
+   *
+   * @covers ::getUserMemberships
+   * @dataProvider membershipDataProvider
+   */
+  public function testGetUserMemberships($user, array $states, $field_name, array $expected) {
+    $result = Og::getUserMemberships($this->users[$user], $states, $field_name);
+
+    // Check that the correct number of results is returned.
+    $this->assertEquals(count($expected), count($result));
+
+    // Inspect the results that were returned.
+    foreach ($result as $key => $membership) {
+      // Check that all result items are OgMembership objects.
+      // @todo Check for \Drupal\og\OgMembershipInterface once OgMembership
+      //   implements it.
+      // @see https://github.com/amitaibu/og/issues/183
+      $this->assertInstanceOf('Drupal\og\Entity\OgMembership', $membership);
+      // Check that the results are keyed by OgMembership ID.
+      $this->assertEquals($membership->id(), $key);
+    }
+
+    // Check that all expected results are returned.
+    foreach ($expected as $expected_group) {
+      $expected_id = $this->groups[$expected_group]->id();
+      foreach ($result as $membership) {
+        if ($membership->getEntityId() === $expected_id) {
+          // Test successful: the expected result was found.
+          continue 2;
+        }
+      }
+      $this->fail("The expected group with ID $expected_id was not found.");
+    }
+  }
+
+  /**
+   * Provides test data to test retrieval of memberships.
+   *
+   * @return array
+   *   An array of test properties. Each property is an indexed array with the
+   *   following items:
+   *   - The key of the user in the $this->users array for which to retrieve
+   *     memberships.
+   *   - An array of membership states to filter on.
+   *   - The field name to filter on.
+   *   - An array containing the expected results to be returned.
+   */
+  public function membershipDataProvider() {
```

Comment:
> Do you think it is overkill?

No - it's exactly the level OG8 should be in. People should feel very comfortable knowing we'll have such a well tested and stable module. Thank you!